src/jit/morph.cpp

```diff
@@ -3187,24 +3187,35 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
         CORINFO_CLASS_HANDLE objClass = NO_CLASS_HANDLE;
         if (isStructArg)
         {
-            // TODO-1stClassStructs: An OBJ node should not be required for lclVars.
-            if (!actualArg->OperIs(GT_OBJ, GT_MKREFANY))
+            objClass = gtGetStructHandle(argx);
+            if (argx->TypeGet() == TYP_STRUCT)
             {
-                BADCODE("illegal argument tree in fgInitArgInfo");
-            }
-            if (actualArg->OperIs(GT_OBJ))
-            {
-                structSize = actualArg->AsObj()->gtBlkSize;
-                objClass   = actualArg->AsObj()->gtClass;
-                assert(structSize == info.compCompHnd->getClassSize(objClass));
+                if (!actualArg->OperIs(GT_OBJ, GT_MKREFANY, GT_LCL_VAR))
+                {
+                    BADCODE("illegal argument tree in fgInitArgInfo");
+                }
+                // For TYP_STRUCT arguments we must have an OBJ, LCL_VAR or MKREFANY
+                switch (argx->OperGet())
+                {
+                    case GT_OBJ:
+                        // Get the size off the OBJ node.
+                        structSize = actualArg->AsObj()->gtBlkSize;
+                        assert(structSize == info.compCompHnd->getClassSize(objClass));
+                        break;
+                    case GT_LCL_VAR:
+                        structSize = lvaGetDesc(argx->AsLclVarCommon())->lvExactSize;
+                        break;
+                    default:
+                        // GT_MKREFANY; perhaps more cases in future.
```

Comment:
I think it would be better to make this an explicit case. And default should probably assert or something. In fact, the BADCODE above should probably be moved here.

Reply:
Sounds good.